mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala

```diff
@@ -137,18 +143,69 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
       assert(model.clusterCenters === model2.clusterCenters)
     }
     val kmeans = new KMeans()
-    testEstimatorAndModelReadWrite(kmeans, dataset, KMeansSuite.allParamSettings, checkModelData)
+    testEstimatorAndModelReadWrite(kmeans, dataset, KMeansSuite.allParamSettings, checkModelData,
+      Map("initialModel" -> (checkModelData _).asInstanceOf[(Any, Any) => Unit]))
+  }
+
+  test("Initialize using a trained model") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(1)
+    val oneIterModel = kmeans.fit(dataset)
+    val twoIterModel = kmeans.copy(ParamMap(ParamPair(kmeans.maxIter, 2))).fit(dataset)
+    val oneMoreIterModel = kmeans.setInitialModel(oneIterModel).fit(dataset)
+
+    assert(oneMoreIterModel.getK === k)
+
+    twoIterModel.clusterCenters.zip(oneMoreIterModel.clusterCenters)
+      .foreach { case (center1, center2) => assert(center1 ~== center2 absTol 1E-8) }
+  }
+
+  test("Initialize using a model with wrong dimension of cluster centers") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(1)
+
+    val wrongDimModel = KMeansSuite.generateRandomKMeansModel(4, k)
+    val wrongDimModelThrown = intercept[IllegalArgumentException] {
+      kmeans.setInitialModel(wrongDimModel).fit(dataset)
+    }
+    assert(wrongDimModelThrown.getMessage.contains("mismatched dimension"))
+  }
+
+  test("Infer K from an initial model") {
+    val kmeans = new KMeans().setK(5)
+    val testNewK = 10
+    val randomModel = KMeansSuite.generateRandomKMeansModel(dim, testNewK)
+    assert(kmeans.setInitialModel(randomModel).getK === testNewK)
+
+    val differentKRandomModel = KMeansSuite.generateRandomKMeansModel(dim, testNewK + 1)
+    assert(kmeans.setInitialModel(differentKRandomModel).getK === testNewK + 1)
+  }
+
+  test("Reset K after setting initial model") {
```

Comment:
If we do change the behavior of just ignoring `k` as noted above, then this test will be invalid. In that case, we should just check that `k` is properly ignored, like:

````scala
  test("Ignore k if initialModel is set") {
    val kmeans = new KMeans()

    val m1 = KMeansSuite.generateRandomKMeansModel(dim, k)
    // ignore k if initialModel is set
    assert(kmeans.setInitialModel(m1).setK(k - 1).getK === k)
    kmeans.clear(kmeans.initialModel)
    // k is not ignored after initialModel is cleared
    assert(kmeans.setK(k - 1).getK === k - 1)
}
````